#!/usr/bin/bash

# run Cufflinks (Trapnell et al. 2012) to assembly the transcripts, from the reads that were previously mapped to the non-coding transcriptome with Hisat2 (Kim et al. 2015), in order to get the .gtf file that is required for FEELnc (Wucher et al. 2017) as an input:
# INPUT: A coordinated bam file and a transcripts annotation file (optional).
# OUTPUT: An assembly in .gtf format.

	cufflinks \
	<COORDINATED_BAM_FILE> \
	-o <OUTPUT_DIRECTORY> \
	-g <ANNOTATION_FILE> \
	-m <AVERAGE_FRAGMENT_LENGTH> \
	-s <STANDARD_DEVIATION>

# <COORDINATED_BAM_FILE>  is the alignment of the reads against the non-coding transcriptome 
# which was generated by the script Non-coding_finder.shnb
# <ANNOTATION_FILE> is the file with the annotated transcripts in gff or gtf format.