#!/usr/bin/env python
# coding: utf-8

# ! conda activate <ACTIVATE_CONDA_ENVIRONMENT>
# 
# run Trinity (Grabherr et al. 2011) in guided mode, in order to assembly *de novo* with genome mapped reads:
# 
#     Trinity \
#     --genome_guided_bam <COORDINATED_BAM_FILE> \
#     --genome_guided_max_intron <MAXIMUM_INTRON_LENGTH> \
#     --max_memory <MAXIMUM_MEMORY> \
#     --CPU <NUMBER_PROCESSORS> \
#     --output <OUTPUT_DIRECTORY> \
#     
#     parameter --genome_guided_bam can be generated by the 
#     aligner of your choice (e.g. Hisat2 (Kim et al. 2015) , 
#     STAR (Dobin et al. 2013)).
#     
# INPUT:Coordinated BAM FILE.
# OUTPUT: *de novo* transcriptome assembly.
